### PR TITLE
Make replay timings accurate

### LINKED ISSUE #104: The replay is slowed down
### Steps to reproduce

1. Attempt a test.
2. Note the time taken.
3. Play the replay.
4. Notice that the time taken is more.

### Expected behaviour
Replay should reflect moderately similar speed.

### Actual behaviour
Replay is slower and takes longer to finish. There's much more time difference between 2 key presses.

### Your proposed fix (if any)
A better way to catch the `ESC` key without timeout would really be helpful. Although, the timeout was introduced to fix https://github.com/Mithil467/mitype/issues/12.
Is there a way to register a signal for the `ESC` key, just like we have for `ctrl-c` in signals.py?

https://github.com/Mithil467/mitype/blob/63cf85813133ef33ff174c50e9041ff801767b8a/mitype/app.py#L519-L525

### PR DESCRIPTION
**Pull request by nkitsaini**
This PR fixes #104 

Tested with python version - 3.10.7

On operating system - Arch Linux

### CODE REVIEW

**Comment by nkitsaini** [comment]
Hi @Mithil467, I'm done with all the changes from my side. Can you please review this when you get some time?

**Comment by Mithil467** [comment]
Thanks for the PR @nkitsaini 🙌

Can you please explain breifly what was the cause of the issue and how does this fix it?

**Comment by nkitsaini** [comment]
@Mithil467 

**TLDR:** Previously we were not accounting for time spent executing the code (displaying the character on screen + checking if `esc` key is pressed). So overtime that time spent will accumulate and create significant delay while replaying. For the new changes I've made the sleep time to be `next_tick - current_time` (next_tick = ideal timestamp for showing keystroke), so the `current_time` here will account for all the time we've previously spent executing code.



**Long Version:**

Here we're replaying the keystrokes. We know the delay between the keystrokes and we just have to replay them.
Let's take following keystrokes as example

```py
[ (1, a), (2, b), (1, c)] # (delay, key)
# a should be shown at 1st second
# b should be shown at 3rd second (1+2)
# c should be shown at 4th second (1+2+1)
```
Previously for each entry we were waiting (`sleep`) for the `delay` seconds and then performed some tasks (display key + see if `esc` is pressed). The issue in this case is that **performing tasks** will consuming some time, and over time it'll build up to create significant delay between replay and actual keystroke timestamps.


Let's examine `c` keystrok from the example. Ideally `c` should been shown exactly at `4 seconds` (1+2+1).

**Previous Version:**
In the previous implementation, the execution would have been something like this:

_For simplicity let's say say `show(key)` = `display key + see if esc pressed + ...`
And Let's say show(key) can take between 0.1-0.2 seconds to execute_

```
# time: operation
0.0: sleep(1) 
1.0: show(a)   (+0.2)
1.2: sleep(2)
3.2: show(b)   (+0.1)
3.3: sleep(1)
4.3: show(c)    
```
Here the `c` was actually shown at `4.3` not `4`.

**Current Version:**
In the current version we don't wait for the delay i.e. we don't execute `sleep(delay)`, instead we calculate at which time a particular keystroke should've been shown. So in case of `c` this value (`next_tick`) will be `1+2+1=4`.
Now when it comes time show `c`, we calculate how far are we from the expected time and sleep for that much duration. 

`next_tick` = when should we show the next keystroke
```
# time: duration
                              next_tick = 0
                              next_tick += 1 (for the `a` keystroke)
0.0: sleep(1)                 next_tick = 1  (sleep_duration = 1-current_time = 1 - 0 = 1)
1.0: show(a)   (+0.2)         next_tick += 2 (for the `b` keystroke)
1.2: sleep(1.8)               next_tick = 3  (sleep_duration = 3-current_time = 3 - 1.2 = 1.8)
3.0: show(b)   (+0.1)         next_tick += 1 (for the `c` keystoke)
3.1: sleep(0.9)               next_tick = 4  (sleep_duration = 4-current_time = 4 - 3.1 = 0.9)
4.0: show(c)
```
This way by accounting for the execution time of `show(a)` we've managed to match the real timestamp while replay.



I know this can be confusing. I hope explanation+example is good enough.
